File: client/json_decoder_handler.go
package client

import (
	"fmt"

	proto "github.com/rgraphql/rgraphql"
)

// jsonDecoderHandler attaches a position in the result tree.
type jsonDecoderHandler struct {
	// resultLock locks the result and calls cb()
	resultLock func(cb func())
	// qnode is the query tree node.
	qnode *qtNode
	// result is the referenced position in the result
	result map[string]interface{}
}

// newJsonDecoderHandler builds a new decoder handler.
func newJsonDecoderHandler(
	resultLock func(cb func()),
	result map[string]interface{},
	qtNode *qtNode,
) *jsonDecoderHandler {
	return &jsonDecoderHandler{
		resultLock: resultLock,
		result:     result,
		qnode:      qtNode,
	}
}

// HandleResultValue handles the next value in the sequence, optionally
// returning a handler for the next value(s) in the sequence.
func (d *jsonDecoderHandler) HandleResultValue(val *proto.RGQLValue) (nextHandler ResultTreeHandler) {
	d.resultLock(func() {
		// Query node, bind new decoder handler to object + field name.
		if qnodeID := val.GetQueryNodeId(); qnodeID != 0 {
			qchild := d.qnode.lookupChildByID(qnodeID)
			if qchild == nil {
				return
			}
			fieldName := qchild.name
			rchild, ok := d.result[fieldName]
			// TODO: We don't know if we need an array or a object here.
			if !ok {
				rchild = make(map[string]interface{})
				d.result[fieldName] = rchild
			}
			rchildMap := rchild.(map[string]interface{})
			nextHandler = newJsonDecoderHandler(
				d.resultLock,
				rchildMap,
				qchild,
			)
			return
		}

		// If array index, return handler bound to the array index.
		if arrayIdx := val.GetArrayIndex(); arrayIdx != 0 {
			fmt.Printf("unhandled array index: %v\n", arrayIdx)
		}
	})
	return
}

var _ ResultTreeHandler = ((*jsonDecoderHandler)(nil))

// arrayChild clones the handler with an array index bound.
/*
func (d *jsonDecoderHandler) arrayChild(idx uint32) *jsonDecoderHandler {
	return &jsonDecoderHandler{
		resultLock: d.resultLock,
		result:     d.result,
		qnode:      d.qnode,
	}
}
*/

File: client/json-decoder-handler.ts
import { PrimitiveValue, UnpackPrimitive } from '../primitive.js'
import { RGQLValue } from '../rgraphql.pb.js'
import { QueryMap, QueryMapElem } from './query-map.js'
import { QueryTreeNode } from './query-tree-node.js'
import { ResultTreeHandler } from './result-tree-handler.js'

// JSONDecoderHandler is a cursor pointing to part of the result.
export class JSONDecoderHandler {
  // value is the current selected value position
  public value?: PrimitiveValue
  // qnode is the attached query tree node.
  public qnode?: QueryTreeNode

  // applyValue applies at the previously selected position
  public applyValue?: (override: boolean, getVal: () => unknown) => PrimitiveValue
  // pendingValue is a pending previous value
  public pendingValue?: RGQLValue

  constructor(
    private queryMap: QueryMap | undefined,
    private valChangedCb: () => void,
  ) {}

  // handleValue is a ResultTreeHandler.
  public handleValue(val: RGQLValue | undefined): ResultTreeHandler {
    let nextHandler = new JSONDecoderHandler(this.queryMap, this.valChangedCb)

    if (val === undefined) {
      if (this.applyValue) {
        this.applyValue(true, () => {
          return undefined
        })
      }
      return null
    }

    if (val.queryNodeId) {
      if (!this.qnode) {
        return null
      }

      const childQnode = this.qnode.lookupChildByID(val.queryNodeId || 0)
      if (!childQnode) {
        return null
      }
      const childFieldName = childQnode.getName()
      let childResultFieldName = childFieldName

      let childQme: QueryMapElem | undefined
      if (this.queryMap) {
        childQme = this.queryMap[val.queryNodeId || 0]
        if (childQme) {
          if (childQme.alias && childQme.alias.length) {
            childResultFieldName = childQme.alias
          }
        }
      }

      if (!childQme) {
        return null
      }

      let nval: PrimitiveValue
      if (this.applyValue) {
        nval = this.applyValue(false, () => {
          return {}
        })
      } else {
        nval = this.value!
      }

      nextHandler.queryMap = childQme.selections
      nextHandler.applyValue = (overrideVal: boolean, getVal: () => unknown) => {
        if (overrideVal || !(childResultFieldName in nval)) {
          const nxval = getVal()
          if (nxval === undefined) {
            delete nval[childResultFieldName]
          } else {
            nval[childResultFieldName] = nxval
          }
          if (this.valChangedCb) {
            this.valChangedCb()
          }
          return nxval as PrimitiveValue
        }
        return nval[childResultFieldName] as PrimitiveValue
      }
      nextHandler.value = nval
      nextHandler.qnode = childQnode
    } else if (val.arrayIndex) {
      let nval: unknown[]
      if (this.applyValue) {
        nval = this.applyValue(false, () => {
          return []
        })
      } else {
        nval = this.value
      }

      const idx = (val.arrayIndex || 1) - 1
      nextHandler.qnode = this.qnode
      nextHandler.applyValue = (override: boolean, getVal: () => unknown) => {
        if (override || !(idx in nval)) {
          const nxval = getVal()
          if (nxval === undefined) {
            // TODO: investigate if this index is consistent.
            nval.splice(idx, 1)
          } else {
            nval[idx] = nxval
          }
          if (this.valChangedCb) {
            this.valChangedCb()
          }
          return nxval
        }
        return nval[idx]
      }
    } else {
      nextHandler = this // eslint-disable-line @typescript-eslint/no-this-alias
    }

    if (val.value) {
      if (nextHandler.applyValue) {
        const unpacked = UnpackPrimitive(val.value)
        nextHandler.applyValue(true, () => {
          return unpacked
        })
      } else {
        // cannot process w/o applyValue function
        return null
      }
    }

    return nextHandler.handleValue.bind(nextHandler)
  }
}
